--- a/compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/selection/MultiWidgetSelectionDelegate.kt
+++ b/compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/selection/MultiWidgetSelectionDelegate.kt
@@ -401,8 +401,8 @@
 // map offsets above/below the text to 0/length respectively
 private fun getOffsetForPosition(position: Offset, textLayoutResult: TextLayoutResult): Int =
     when {
-        position.y <= 0f -> 0
-        position.y >= textLayoutResult.multiParagraph.height ->
+        position.y < 0f -> 0
+        position.y > textLayoutResult.size.height ->
             textLayoutResult.layoutInput.text.length
         else -> textLayoutResult.getOffsetForPosition(position)
     }
--- a/compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.kt
+++ b/compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.kt
@@ -1444,7 +1444,12 @@
         val needToRecomputeSelection =
             selectionLayout.shouldRecomputeSelection(previousSelectionLayout)
         if (needToRecomputeSelection) {
-            val newSelection = adjustment.adjust(selectionLayout)
+            var newSelection = adjustment.adjust(selectionLayout)
+            if (newSelection.start.selectableId == newSelection.end.selectableId &&
+                newSelection.handlesCrossed != (newSelection.start.offset > newSelection.end.offset)
+            ) {
+                newSelection = newSelection.copy(handlesCrossed = !newSelection.handlesCrossed)
+            }
             if (newSelection != selection) {
                 if (shouldPerformHaptics()) {
                     hapticFeedBack?.performHapticFeedback(HapticFeedbackType.TextHandleMove)
